### DISCUSSION TITLE
Cache a route after user visits it with nuxt hybrid rendering and ssr

### DISCUSSION THREAD

**Discussion by dbzx10299** [vite-pwa/vite-plugin-pwa#734] [Q&A] [answered]
When using nuxt3 with `ssr` how can we cache a route after the user visits it. It isn‘t working for me unless I use pre rendering. I have a good idea why this is happening, although I don't know how to handle this. When the `npm run build` command is run, the output is `dist/_nuxt/b-__gekbe.js` and a whole bunch of other files that have some hash. They are indeed being cached but the page does not seem to be cached because after I visit the page, if I go offline then re-visit, I get a blank screen despite the js payload being cached.

I have a project structure like this
```txt
pages/
└── products/
    └── [handle].vue
nuxt.config.ts
```

and in my `nuxt.config.ts` I have something like this
```js
export default defineNuxtConfig({
  devtools: { 
    enabled: false 
  },
  css: [
    '@/assets/main.css'
  ],
  modules: [
    '@vite-pwa/nuxt'
  ],
  pwa: {
    manifest: {
      name: 'foo',
      short_name: 'foo',
      background_color: '#f0f0f0',
      theme_color: '#f0f0f0',
      orientation: 'portrait',
      display: 'standalone',
      icons: [
        {
          src: 'pwa-192x192.png',
          sizes: '192x192',
          type: 'image/png'
        },
        {
          src: 'pwa-512x512.png',
          sizes: '512x512',
          type: 'image/png'
        }
      ]
    },
    registerType: 'autoUpdate',
    workbox: {
      navigateFallback: undefined,
      globPatterns: [
        "**/*.{js,css,html,png,jpg,jpeg,svg,webp,webm,woff2,woff,ttf,eot,webmanifest}"
      ],
      runtimeCaching: [
        {
          urlPattern: /^https:\/\/cdn\.shopify\.com\/.*/i,
          handler: 'CacheFirst',
          options: {
            cacheName: 'shopify-file-cache',
            expiration: {
              maxEntries: 10,
              maxAgeSeconds: 60 * 60 * 24 * 365 // <== 365 days
            },
            cacheableResponse: {
              statuses: [0, 200]
            }
          }
        }
      ]
    }
  },
  nitro: {
    preset: 'netlify'
  },
  routeRules: {
    '/products/**': {
      headers: {
        'netlify-cache-tag': 'products',
        'cache-control': 'public, max-age=3600, must-revalidate',
        'netlify-cdn-cache-control': 'public, s-maxage=300, stale-while, revalidate=604800'
      }
    }
  },
  compatibilityDate: '2024-07-06'
})
```

**Comment by userquin** [comment]
https://github.com/vite-pwa/sveltekit/issues/65#issuecomment-1734213190

**Reply by dbzx10299** [reply]
I did take a look at the link and played around with this the past hour and I still can't quite get it to work. I updated the `nuxt.config.ts` file to this and I still see a blank screen if I go offline and navigate back to the product page.

I have a route at `/products/[handle].vue` so going to `home /` `=>` going offline `=>` back to `/products/shirt1` while offline gives a white screen. One thing I am not sure about is although this is on nuxt3 and is `ssr` navigating around with `<NuxtLink to="..."` causes behavior like a `SPA`

```js
export default defineNuxtConfig({
  modules: [
    '@vite-pwa/nuxt'
  ],
  pwa: {
    manifest: {
      name: 'foo',
      short_name: 'foo',
      background_color: '#f0f0f0',
      theme_color: '#f0f0f0',
      orientation: 'portrait',
      display: 'standalone',
      icons: [
        {
          src: 'pwa-192x192.png',
          sizes: '192x192',
          type: 'image/png'
        },
        {
          src: 'pwa-512x512.png',
          sizes: '512x512',
          type: 'image/png'
        }
      ]
    },
    registerType: 'autoUpdate',
    workbox: {
      navigateFallback: undefined,
      navigateFallbackDenylist: [/^\/products\/.*/],
      globPatterns: [
        "**/*.{js,css,html,png,jpg,jpeg,svg,webp,webm,woff2,woff,ttf,eot,webmanifest}"
      ],
      runtimeCaching: [
        {
          urlPattern: ({ url, sameOrigin }) => sameOrigin && url.pathname.match(/^\/products\/.*/),
          handler: 'NetworkFirst',
          options: {
            cacheName: 'ssr-pages-cache',
            cacheableResponse: {
              statuses: [200]
            },
            matchOptions: {
              ignoreVary: true,
              ignoreSearch: true
            },
            plugins: [{
              cacheWillUpdate: async ({ response }) => response.status === 200 ? response : null
            }]
          },
        }
      ]
    }
  }
})
````

**Reply by userquin** [reply]
Try removing trailing `.*` from `navigateFallbackDenylist` and `urlPattern` from runtime caching.

You can try `nuxt build && node ...` in your local enabling development mode in pwa options, you will see workbox modules logs:
```js
pwa: {
  mode: 'development',
 ... // other options
}
```

**Reply by dbzx10299** [reply]
This still is not working, I am going to have to deep dive into this, I think this might be related to `payloadExtraction` needing to be set to `true`.

**Reply by dbzx10299** [reply]
I just updated it again to this since the nuxt 3 framework docs on [payload extraction](https://vite-pwa-org.netlify.app/frameworks/nuxt#payload-extraction) state

> When you enable the experimental payloadExtraction flag in your Nuxt configuration file, @vite-pwa/nuxt will add **/_payload.json to the globPatterns array inside workbox or injectManifest option, depending on the configured strategy.

```js
export default defineNuxtConfig({
 
  modules: [
    '@vite-pwa/nuxt'
  ],
  pwa: {
    strategies: 'generateSW',
    manifest: {
      name: 'Dolce & Gabbana',
      short_name: 'Dolce & Gabbana',
      background_color: '#f0f0f0',
      theme_color: '#f0f0f0',
      orientation: 'portrait',
      display: 'standalone',
      icons: [
        {
          src: 'pwa-192x192.png',
          sizes: '192x192',
          type: 'image/png'
        },
        {
          src: 'pwa-512x512.png',
          sizes: '512x512',
          type: 'image/png'
        }
      ]
    },
    registerType: 'autoUpdate',
    workbox: {
      navigateFallback: undefined,
      globPatterns: [
        "**/*.{js,css,html,png,jpg,jpeg,svg,webp,webm,woff2,woff,ttf,eot,webmanifest}",
        "_nuxt/builds/**/*.json",
        "**/_payload.json"
      ]
    }
  }
})
```

I still am getting the white screen for some reason. I'm running out of ideas here what the culprit could be

**Reply by userquin** [reply]
can you prepare small reproduction? I'll check it in a few hours...

**Reply by dbzx10299** [reply]
I have some things like api keys that need to be private, I will try to make a reproduction but it will be a bit of a pain since I‘m pulling data into the site from other sources.

**Reply by userquin** [reply]
I'm preparing small reproduction...

**Reply by dbzx10299** [reply]
I am using the shopify storefront api, I have to test this but I am pulling external resources in with the image cdn, I believe I may need to do the same with the graphql endpoint

**Reply by userquin** [reply]
![imagen](https://github.com/user-attachments/assets/50a224a1-b39f-4d6d-8254-5256aa1f9ad5)

**Reply by userquin** [reply]
Since nuxt use hybrid navigation mode, there is no api call on first request, on next requests the `useFetch` will use th eapi call, and so we need to include in denylist ssr pages and api calls via a custom handler for:
- server pages (dynamic ones or not)
- api calls

In the playground I'm preparing there are 2 server pages and 2 api calls:
- `/products` and `/products/[id]`  pages and
- `/api/products` and `/api/products/[id]` api calls

**Reply by userquin** [reply]
The playground will use custom service worker.

**Reply by userquin** [reply, marked as answer]
https://github.com/userquin/vite-pwa-nuxt-ssr-pages

**Reply by userquin** [reply]
You can also include 404 and 410 statuses to allow caching server pages and api calls for missing entries... otherwise redirected to offline page on page refresh

**Reply by dbzx10299** [reply]
Wow thanks a lot for taking the time to create a repo for this! I will study this and see how you got this to work, this would be useful in the docs as well. I take it that for something like this is must be done with `injectManifest` and using the pwa config in `nuxt.config.ts` is not granular enough.

**Reply by userquin** [reply]
@dbzx10299 check last commit about  `cachedResponseWillBeUsed`, we can sotre the cached response in the state and return that response, now any request when offline present in the cache will work

**Reply by dbzx10299** [reply]
Got it. One thing just so I'm clear, for `ssr` specifically, do we need to use `injectManifest` and bypass the `pwa.workbox.runtimeCaching[{/*...*/}]`. I have been reading the docs and I see that the injectManifest seems quite a bit more complex but also more powerful and allows for more fine tuning.

**Reply by userquin** [reply]
> Wow thanks a lot for taking the time to create a repo for this! I will study this and see how you got this to work, this would be useful in the docs as well. I take it that for something like this is must be done with `injectManifest` and using the pwa config in `nuxt.config.ts` is not granular enough.

PR welcome to docs repo

**Reply by userquin** [reply]
> Got it. One thing just so I'm clear, for `ssr` specifically, do we need to use `injectManifest` and bypass the `pwa.workbox.runtimeCaching[{/*...*/}]`. I have been reading the docs and I see that the injectManifest seems quite a bit more complex but also more powerful and allows for more fine tuning.

`generateSW` will build the custom sw for you using the same modules (check the static imports, workbox-build will generate the same code), the custom sw is about having the full control

**Reply by userquin** [reply]
One last thing about workbox plugins, we're using cacheable and expiration, there are a few more and also some internal ones (we're only using 2 hooks):

![imagen](https://github.com/user-attachments/assets/f6102a8a-6041-4ac8-980b-5fb4fcc3c0c5)

**Reply by userquin** [reply]
Nuxt 3 v4-compat version: https://github.com/userquin/vite-pwa-nuxt4-ssr-pages (a few changes in project layout and server api imports): created from this PR https://github.com/vite-pwa/create-pwa/pull/18

**Reply by userquin** [reply]
Added `ignoreURLParametersMatching`  to `precacheAndRoute` handler in both repos, nuxt requesting json payload with the version as query param. Check last commit.
